### PR TITLE
Support long decimal in prefixsort

### PR DESCRIPTION
**Pull request by jinchengchenghh**
The performance improve from 1.5 ~2.8 times vs stdsort. The larger the amount of data, the greater the optimization。

```
============================================================================
[...]ec/benchmarks/PrefixSortBenchmark.cpp     relative  time/iter   iters/s
============================================================================
StdSort_no-payloads_1_hugeint_1k                          157.05ns     6.37M
PrefixSort                                      159.96%    98.18ns    10.19M
StdSort_no-payloads_2_hugeint_1k                          229.25ns     4.36M
PrefixSort                                      187.99%   121.95ns     8.20M
StdSort_no-payloads_3_hugeint_1k                          300.49ns     3.33M
PrefixSort                                      214.49%   140.09ns     7.14M
StdSort_no-payloads_4_hugeint_1k                          334.26ns     2.99M
PrefixSort                                      177.37%   188.45ns     5.31M
StdSort_no-payloads_1_hugeint_10k                         223.07ns     4.48M
PrefixSort                                      179.30%   124.41ns     8.04M
StdSort_no-payloads_2_hugeint_10k                         299.86ns     3.33M
PrefixSort                                      201.84%   148.57ns     6.73M
StdSort_no-payloads_3_hugeint_10k                         389.77ns     2.57M
PrefixSort                                      233.71%   166.77ns     6.00M
StdSort_no-payloads_4_hugeint_10k                         473.84ns     2.11M
PrefixSort                                      219.03%   216.34ns     4.62M
StdSort_no-payloads_1_hugeint_100k                        288.05ns     3.47M
PrefixSort                                      192.22%   149.86ns     6.67M
StdSort_no-payloads_2_hugeint_100k                        403.63ns     2.48M
PrefixSort                                      225.74%   178.80ns     5.59M
StdSort_no-payloads_3_hugeint_100k                        511.05ns     1.96M
PrefixSort                                      257.16%   198.73ns     5.03M
StdSort_no-payloads_4_hugeint_100k                        606.13ns     1.65M
PrefixSort                                      237.37%   255.36ns     3.92M
StdSort_no-payloads_1_hugeint_1000k                       388.12ns     2.58M
PrefixSort                                      216.15%   179.56ns     5.57M
StdSort_no-payloads_2_hugeint_1000k                       546.45ns     1.83M
PrefixSort                                      246.45%   221.73ns     4.51M
StdSort_no-payloads_3_hugeint_1000k                       716.56ns     1.40M
PrefixSort                                      274.69%   260.86ns     3.83M
StdSort_no-payloads_4_hugeint_1000k                       952.00ns     1.05M
PrefixSort                                      283.40%   335.92ns     2.98M
```

### CODE REVIEW

**Comment by jinchengchenghh** [comment]
@skadilover Can you help review this PR? Thanks!

File: velox/exec/benchmarks/PrefixSortBenchmark.cpp [hunk 1]
```diff
@@ -298,6 +298,35 @@ class PrefixSortBenchmark {
         "no-payloads", "varchar", batchSizes, rowTypes, numKeys, iterations);
   }
 
+  void hugeInt() {
+    const auto iterations = 10;
+    const std::vector<vector_size_t> batchSizes = {
+        1'000, 10'000, 100'000, 1'000'000};
```
Comment (skadilover) [inline velox/exec/benchmarks/PrefixSortBenchmark.cpp:304]: Could you share the benchmark result ？

File: velox/exec/PrefixSort.cpp [hunk 2]
```diff
@@ -120,7 +125,7 @@ PrefixSortLayout PrefixSortLayout::makeSortLayout(
   uint32_t numNormalizedKeys = 0;
   const uint32_t numKeys = types.size();
   std::vector<uint32_t> prefixOffsets;
-  std::vector<PrefixSortEncoder> encoders;
+  std::vector<std::shared_ptr<PrefixSortEncoder>> encoders;
```
Comment (skadilover) [inline velox/exec/PrefixSort.cpp:128]: Could you explain why we are making this change?

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 3]
```diff
@@ -46,10 +51,18 @@ class PrefixSortEncoder {
   FOLLY_ALWAYS_INLINE void encode(std::optional<T> value, char* dest) const {
     if (value.has_value()) {
       dest[0] = nullsFirst_ ? 1 : 0;
-      encodeNoNulls(value.value(), dest + 1);
```
Comment (skadilover) [inline velox/exec/prefixsort/PrefixSortEncoder.h:49]: I don’t prefer to add an if statement to the hot code. Can we use encodeNoNulls for compatibility processing? If not, can you try refactoring this method?

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 4]
```diff
@@ -228,4 +256,48 @@ FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
...
+  PrefixSortHugeIntEncoder(
+      bool ascending,
+      bool nullsFirst,
+      int precision,
+      int scale)
+      : PrefixSortEncoder(ascending, nullsFirst),
+        precision_(precision),
+        scale_(scale){};
+
+  void encodeHugeInt(int128_t value, char* dst) const override {
```
Comment (skadilover) [inline velox/exec/prefixsort/PrefixSortEncoder.h:270]: Can we use template methods in PrefixSortEncoder to achieve this functionality?

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 5]
```diff
@@ -228,4 +256,48 @@ FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
...
+    if (precision_ - scale_ <= ShortDecimalType::kMaxPrecision) {
```
Comment (skadilover) [inline velox/exec/prefixsort/PrefixSortEncoder.h:271]: Can we provide separate implementations for different precisions?

**Comment by skadilover** [comment]
@jinchengchenghh Thank you very much for extending the implementation of prefix-sort.

Matainer may need to continue to review it in the future. cc @mbasmanova

**Comment by mbasmanova** [comment]
@jinchengchenghh There are CI failures. Would you take a look? 

@skadilover Would you help take a first pass?

**Comment by skadilover** [comment]
> @jinchengchenghh There are CI failures. Would you take a look?
> 
> @skadilover Would you help take a first pass?

ok~

File: velox/exec/PrefixSort.cpp (hunk 2 above)
Comment (jinchengchenghh) [reply velox/exec/PrefixSort.cpp:128]: Because for hugeint, we need reserve precision and scale, so it extends PrefixSortEncoder, to save it, we need to make it shared_ptr

**Comment by jinchengchenghh** [comment]
============================================================================
[...]ec/benchmarks/PrefixSortBenchmark.cpp     relative  time/iter   iters/s
============================================================================
StdSort_no-payloads_1_hugeint_1k                          900.54ns     1.11M
PrefixSort                                      121.58%   740.67ns     1.35M
StdSort_no-payloads_2_hugeint_1k                            1.51us   661.06K
PrefixSort                                      144.97%     1.04us   958.33K
StdSort_no-payloads_3_hugeint_1k                            1.90us   525.68K
PrefixSort                                      149.71%     1.27us   787.00K
StdSort_no-payloads_4_hugeint_1k                            2.40us   415.86K
PrefixSort                                      162.80%     1.48us   677.00K
StdSort_no-payloads_1_hugeint_10k                           1.16us   859.82K
PrefixSort                                      132.32%   878.99ns     1.14M
StdSort_no-payloads_2_hugeint_10k                           2.00us   500.87K
PrefixSort                                      159.51%     1.25us   798.92K
StdSort_no-payloads_3_hugeint_10k                           2.67us   373.98K
PrefixSort                                      175.94%     1.52us   657.99K
StdSort_no-payloads_4_hugeint_10k                           3.46us   288.89K
PrefixSort                                      198.48%     1.74us   573.39K
StdSort_no-payloads_1_hugeint_100k                          1.54us   649.38K
PrefixSort                                      146.99%     1.05us   954.51K
StdSort_no-payloads_2_hugeint_100k                          2.52us   396.33K
PrefixSort                                      175.58%     1.44us   695.89K
StdSort_no-payloads_3_hugeint_100k                          3.45us   289.83K
PrefixSort                                      194.63%     1.77us   564.11K
StdSort_no-payloads_4_hugeint_100k                          4.33us   231.20K
PrefixSort                                      209.31%     2.07us   483.92K
StdSort_no-payloads_1_hugeint_1000k                         1.95us   512.46K
PrefixSort                                      161.34%     1.21us   826.80K
StdSort_no-payloads_2_hugeint_1000k                         3.06us   327.16K
PrefixSort                                      181.59%     1.68us   594.08K
StdSort_no-payloads_3_hugeint_1000k                         4.26us   234.60K
PrefixSort                                      210.04%     2.03us   492.76K
StdSort_no-payloads_4_hugeint_1000k                         5.73us   174.57K
PrefixSort                                      243.59%     2.35us   425.23K
StdSort_no-payloads-prefix-int128_1_hugeint_1k            869.70ns     1.15M
PrefixSort                                      125.82%   691.21ns     1.45M
StdSort_no-payloads-prefix-int128_2_hugeint_1k              1.45us   688.01K
PrefixSort                                      138.78%     1.05us   954.79K
StdSort_no-payloads-prefix-int128_3_hugeint_1k              2.05us   488.96K
PrefixSort                                      152.03%     1.35us   743.37K
StdSort_no-payloads-prefix-int128_4_hugeint_1k              2.36us   424.59K
PrefixSort                                      137.00%     1.72us   581.67K
StdSort_no-payloads-prefix-int128_1_hugeint_10k             1.20us   835.23K
PrefixSort                                      139.99%   855.25ns     1.17M
StdSort_no-payloads-prefix-int128_2_hugeint_10k             1.93us   517.34K
PrefixSort                                      152.48%     1.27us   788.86K
StdSort_no-payloads-prefix-int128_3_hugeint_10k             2.68us   373.61K
PrefixSort                                      167.17%     1.60us   624.56K
StdSort_no-payloads-prefix-int128_4_hugeint_10k             3.59us   278.52K
PrefixSort                                      174.38%     2.06us   485.69K
StdSort_no-payloads-prefix-int128_1_hugeint_100             1.56us   639.74K
PrefixSort                                      152.50%     1.03us   975.58K
StdSort_no-payloads-prefix-int128_2_hugeint_100             2.52us   396.36K
PrefixSort                                      167.22%     1.51us   662.79K
StdSort_no-payloads-prefix-int128_3_hugeint_100             3.41us   293.13K
PrefixSort                                      180.21%     1.89us   528.25K
StdSort_no-payloads-prefix-int128_4_hugeint_100             4.29us   232.86K
PrefixSort                                      177.64%     2.42us   413.65K
StdSort_no-payloads-prefix-int128_1_hugeint_100             1.96us   510.85K
PrefixSort                                      161.67%     1.21us   825.88K
StdSort_no-payloads-prefix-int128_2_hugeint_100             3.10us   322.94K
PrefixSort                                      174.73%     1.77us   564.29K
StdSort_no-payloads-prefix-int128_3_hugeint_100             4.24us   235.73K
PrefixSort                                      192.66%     2.20us   454.16K
StdSort_no-payloads-prefix-int128_4_hugeint_100             5.99us   166.89K
PrefixSort                                      210.47%     2.85us   351.25K

**Comment by jinchengchenghh** [comment]
```
============================================================================
[...]ec/benchmarks/PrefixSortBenchmark.cpp     relative  time/iter   iters/s
============================================================================
StdSort_no-payloads_1_hugeint_1k                          159.96ns     6.25M
PrefixSort                                      111.31%   143.71ns     6.96M
StdSort_no-payloads_2_hugeint_1k                          230.78ns     4.33M
PrefixSort                                      133.04%   173.47ns     5.76M
StdSort_no-payloads_3_hugeint_1k                          275.46ns     3.63M
PrefixSort                                      138.86%   198.37ns     5.04M
StdSort_no-payloads_4_hugeint_1k                          329.72ns     3.03M
PrefixSort                                      165.85%   198.81ns     5.03M
StdSort_no-payloads_1_hugeint_10k                         217.04ns     4.61M
PrefixSort                                      127.67%   170.00ns     5.88M
StdSort_no-payloads_2_hugeint_10k                         304.38ns     3.29M
PrefixSort                                      150.54%   202.20ns     4.95M
StdSort_no-payloads_3_hugeint_10k                         380.78ns     2.63M
PrefixSort                                      165.46%   230.13ns     4.35M
StdSort_no-payloads_4_hugeint_10k                         455.95ns     2.19M
PrefixSort                                      201.34%   226.46ns     4.42M
StdSort_no-payloads_1_hugeint_100k                        293.45ns     3.41M
PrefixSort                                      148.18%   198.03ns     5.05M
StdSort_no-payloads_2_hugeint_100k                        413.40ns     2.42M
PrefixSort                                      174.98%   236.26ns     4.23M
StdSort_no-payloads_3_hugeint_100k                        528.00ns     1.89M
PrefixSort                                      193.18%   273.32ns     3.66M
StdSort_no-payloads_4_hugeint_100k                        613.82ns     1.63M
PrefixSort                                      229.48%   267.49ns     3.74M
StdSort_no-payloads_1_hugeint_1000k                       398.68ns     2.51M
PrefixSort                                      169.93%   234.62ns     4.26M
StdSort_no-payloads_2_hugeint_1000k                       551.28ns     1.81M
PrefixSort                                      190.58%   289.27ns     3.46M
StdSort_no-payloads_3_hugeint_1000k                       745.30ns     1.34M
PrefixSort                                      230.99%   322.66ns     3.10M
StdSort_no-payloads_4_hugeint_1000k                       960.51ns     1.04M
PrefixSort                                      297.72%   322.62ns     3.10M
StdSort_no-payloads-prefix-int128_1_hugeint_1k            148.83ns     6.72M
PrefixSort                                      129.79%   114.66ns     8.72M
StdSort_no-payloads-prefix-int128_2_hugeint_1k            219.16ns     4.56M
PrefixSort                                      150.78%   145.35ns     6.88M
StdSort_no-payloads-prefix-int128_3_hugeint_1k            292.17ns     3.42M
PrefixSort                                      174.02%   167.89ns     5.96M
StdSort_no-payloads-prefix-int128_4_hugeint_1k            323.80ns     3.09M
PrefixSort                                      147.78%   219.11ns     4.56M
StdSort_no-payloads-prefix-int128_1_hugeint_10k           219.32ns     4.56M
PrefixSort                                      157.68%   139.09ns     7.19M
StdSort_no-payloads-prefix-int128_2_hugeint_10k           300.22ns     3.33M
PrefixSort                                      171.60%   174.95ns     5.72M
StdSort_no-payloads-prefix-int128_3_hugeint_10k           388.86ns     2.57M
PrefixSort                                      195.57%   198.83ns     5.03M
StdSort_no-payloads-prefix-int128_4_hugeint_10k           469.14ns     2.13M
PrefixSort                                      184.65%   254.07ns     3.94M
StdSort_no-payloads-prefix-int128_1_hugeint_100           287.41ns     3.48M
PrefixSort                                      172.37%   166.74ns     6.00M
StdSort_no-payloads-prefix-int128_2_hugeint_100           404.13ns     2.47M
PrefixSort                                      191.45%   211.09ns     4.74M
StdSort_no-payloads-prefix-int128_3_hugeint_100           516.65ns     1.94M
PrefixSort                                      216.93%   238.16ns     4.20M
StdSort_no-payloads-prefix-int128_4_hugeint_100           608.44ns     1.64M
PrefixSort                                      196.04%   310.37ns     3.22M
StdSort_no-payloads-prefix-int128_1_hugeint_100           393.71ns     2.54M
PrefixSort                                      189.50%   207.76ns     4.81M
StdSort_no-payloads-prefix-int128_2_hugeint_100           562.31ns     1.78M
PrefixSort                                      214.66%   261.95ns     3.82M
StdSort_no-payloads-prefix-int128_3_hugeint_100           739.96ns     1.35M
PrefixSort                                      244.99%   302.04ns     3.31M
StdSort_no-payloads-prefix-int128_4_hugeint_100           986.38ns     1.01M
PrefixSort                                      251.19%   392.69ns     2.55M
```

File: velox/exec/PrefixSort.cpp [hunk 6]
```diff
@@ -37,9 +37,15 @@ FOLLY_ALWAYS_INLINE void encodeRowColumn(
   if (RowContainer::isNullAt(row, rowColumn.nullByte(), rowColumn.nullMask())) {
     value = std::nullopt;
   } else {
-    value = *(reinterpret_cast<T*>(row + rowColumn.offset()));
+    if constexpr (std::is_same_v<T, int128_t>) {
+      int128_t newVal;
+      memcpy(&newVal, row + rowColumn.offset(), sizeof(int128_t));
```
Comment (jinchengchenghh) [inline velox/exec/PrefixSort.cpp:42]: Because the row address is not 16B alignment, so we need to copy it

**Comment by jinchengchenghh** [comment]
Refactored, can you help review again? Thanks! @skadilover

File: velox/exec/PrefixSort.cpp (hunk 6 above)
Comment (jinchengchenghh) [reply velox/exec/PrefixSort.cpp:42]: Otherwise the release version will core dump

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 7]
```diff
@@ -228,4 +237,126 @@ FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
...
+  PrefixSortHugeIntEncoder(
+      bool ascending,
+      bool nullsFirst,
+      int precision,
+      int scale)
+      : PrefixSortEncoder(ascending, nullsFirst),
+        precision_(precision),
+        scale_(scale){};
+
+  FOLLY_ALWAYS_INLINE void encode(std::optional<int128_t> value, char* dest)
```
Comment (skadilover) [inline velox/exec/prefixsort/PrefixSortEncoder.h:302]: encode() methhod is  "hot-code", please not add virtual call.

Could you try using this approach to handle different types of logic?
```
void RowContainer::store(
    const DecodedVector& decoded,
    vector_size_t index,
    char* row,
    int32_t column) {
  auto numKeys = keyTypes_.size();
  if (column < numKeys && !nullableKeys_) {
    VELOX_DYNAMIC_TYPE_DISPATCH(
        storeNoNulls,
        typeKinds_[column],
        decoded,
        index,
        row,
        offsets_[column]);
  } else {
    VELOX_DCHECK(column < keyTypes_.size() || accumulators_.empty());
    auto rowColumn = rowColumns_[column];
    VELOX_DYNAMIC_TYPE_DISPATCH_ALL(
        storeWithNulls,
        typeKinds_[column],
        decoded,
        index,
        row,
        rowColumn.offset(),
        rowColumn.nullByte(),
        rowColumn.nullMask());
  }
}
```

Comment (jinchengchenghh) [reply velox/exec/prefixsort/PrefixSortEncoder.h:302]: This virtual method is just for int128_t, other type is not virtual
https://github.com/facebookincubator/velox/pull/10385/files#diff-68588996b759bc7bd50b7eff21fcfc2b4ec7b67d234783b348a8af24cbd38a2fR59

Comment (jinchengchenghh) [reply velox/exec/prefixsort/PrefixSortEncoder.h:302]: encoders is in SortLayout, so we cannot use non-virtual function for int128, we should change `extractRowColumnToPrefix` to `VELOX_DYNAMIC_TYPE_DISPATCH_ALL`

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 8]
```diff
@@ -53,6 +56,10 @@ class PrefixSortEncoder {
     }
   }
 
+  virtual void encode(std::optional<int128_t> value, char* dest) const {
```
Comment (skadilover) [inline velox/exec/prefixsort/PrefixSortEncoder.h:59]: I think we can add a method encodeForDecimal here which can have new parameters precision & scale
For varchar and varbinary we can also add here.

File: velox/exec/prefixsort/PrefixSortEncoder.h (hunk 7 above)
Comment (skadilover) [reply velox/exec/prefixsort/PrefixSortEncoder.h:302]: Just an example, the "type dispatch code" is done at extractRowColumnToPrefix already, I think we just need to add    
'case branch' for new types.

File: velox/exec/prefixsort/PrefixSortEncoder.h (hunk 8 above)
Comment (jinchengchenghh) [reply velox/exec/prefixsort/PrefixSortEncoder.h:59]: Then we need to check if the value with this datatype can cast to int for each value, if cast to int, we should get the new scale for each value, do we accept this performance issue?

Comment (skadilover) [reply velox/exec/prefixsort/PrefixSortEncoder.h:59]: I think I understand your intention. I think the difference between this and other types of type dispach is that an additional judgment of precision and scale needs to be added. This cost needs to be weighed against the cost of a virtual call. I think if there is no complicated To calculate expressions, it would be better to add a judgment to type dispatch.

I no longer insist on my opinion.

Comment (jinchengchenghh) [reply velox/exec/prefixsort/PrefixSortEncoder.h:59]: type dispatch `VELOX_DYNAMIC_TYPE_DISPATCH_ALL` uses switch case for different types, I think the expense is bigger than virtual function. What do you think about this? @mbasmanova

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 9]
```diff
@@ -228,4 +237,126 @@ FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
...
+      dest[0] = nullsFirst_ ? 1 : 0;
+      int64_t newValue;
+      auto status = DecimalUtil::rescaleWithRoundUp<int128_t, int64_t>(
+          value.value(),
+          precision_,
+          scale_,
+          newPrecision_,
+          newScale_,
+          newValue);
+      if (FOLLY_UNLIKELY(!status.ok())) {
```
Comment (skadilover) [inline velox/exec/prefixsort/PrefixSortEncoder.h:266]: Why here we can consider status not ok as an expected behavior ？
Can you add a comment here to explain?

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 10]
```diff
@@ -228,4 +237,126 @@ FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
...
+      : PrefixSortEncoder(ascending, nullsFirst),
+        precision_(precision),
+        scale_(scale){};
+
+  FOLLY_ALWAYS_INLINE void encode(std::optional<int128_t> value, char* dest)
+      const override {
+    if (value.has_value()) {
+      dest[0] = nullsFirst_ ? 1 : 0;
+      auto val = value.value();
+      encodeNoNulls<int64_t>(HugeInt::upper(val), dest + 1);
```
Comment (skadilover) [inline velox/exec/prefixsort/PrefixSortEncoder.h:307]: I think we should not make semantic changes here in sort？Can you prove that this is OK under any circumstances? Or is there any relevant document reference?

**Review by skadilover — COMMENTED** [review]
Excluding virtual method questions, two new questions have been added.

File: velox/exec/prefixsort/PrefixSortEncoder.h (hunk 10 above)
Comment (jinchengchenghh) [reply velox/exec/prefixsort/PrefixSortEncoder.h:307]: https://github.com/duckdb/duckdb/blob/main/src/include/duckdb/common/radix.hpp#L225 duckdb implementation

File: velox/exec/prefixsort/PrefixSortEncoder.h (hunk 9 above)
Comment (jinchengchenghh) [reply velox/exec/prefixsort/PrefixSortEncoder.h:266]: Spark implementation https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala#L172

Background
https://issues.apache.org/jira/browse/SPARK-40089

Comment (skadilover) [reply velox/exec/prefixsort/PrefixSortEncoder.h:266]: It would be nice to add an ut for these case, or do I miss it ?

**Comment by jinchengchenghh** [comment]
Do you have further comments? @skadilover

File: velox/exec/prefixsort/tests/PrefixEncoderTest.cpp [hunk 11]
```diff
@@ -350,18 +356,37 @@ TEST_F(PrefixEncoderTest, encodeHugeInt) {
   auto size = PrefixSortEncoderFactory::encodedSize(*type1);
```
Comment (skadilover) [inline velox/exec/prefixsort/tests/PrefixEncoderTest.cpp:356]: It takes a lot of effort to understand the intention here.
Please add some necessary comments to explain how ut ensures that all cases are covered（What scenarios are data patterns used to cover?）, especially the special paths introduced from spark & ​​duckDB mentioned above.

**Review by skadilover — COMMENTED** [review]
@jinchengchenghh 
two comments here～

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 12]
```diff
@@ -315,10 +309,6 @@ class PrefixSortHugeIntEncoder : public PrefixSortEncoder {
   static int32_t encodedSize() {
     return 17;
   }
-
- private:
-  const int precision_;
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/exec/prefixsort/PrefixSortEncoder.h]: We just move these 2 for LongDecimalToHugeIntEncoder, but LongDecimalCastToIntEncoder still needs this information

File: velox/exec/prefixsort/tests/PrefixEncoderTest.cpp (hunk 11 above)
Comment (jinchengchenghh) [reply velox/exec/prefixsort/tests/PrefixEncoderTest.cpp:356]: I add the new decimal encodedSize, so we need to test for decimal.
Other logic for `PrefixSortEncoder` should be tested before.

Comment (skadilover) [reply velox/exec/prefixsort/tests/PrefixEncoderTest.cpp:356]: So currently we haven't figured out whether the two code branches mentioned above are covered by ut?

Comment (jinchengchenghh) [reply velox/exec/prefixsort/tests/PrefixEncoderTest.cpp:356]: It has covered. I add the tests to test decimal can cast to int and hugeint itself

Comment (skadilover) [reply velox/exec/prefixsort/tests/PrefixEncoderTest.cpp:356]: I think ut should be added here separately for them

Comment (jinchengchenghh) [reply velox/exec/prefixsort/tests/PrefixEncoderTest.cpp:356]: https://github.com/facebookincubator/velox/pull/10385/files#diff-669e68649d93c35ad69bd0724e11e6b181021954e5483076ad88604ee67d5fcbR291
hugeIntEncoder1_ tests cast to int case, and hugeIntEncoder2_ tests for hugeint

Comment (jinchengchenghh) [reply velox/exec/prefixsort/tests/PrefixEncoderTest.cpp:356]: hugeIntEncoder1_  https://github.com/facebookincubator/velox/pull/10385/files#diff-669e68649d93c35ad69bd0724e11e6b181021954e5483076ad88604ee67d5fcbR359 
hugeIntEncoder2_  https://github.com/facebookincubator/velox/pull/10385/files#diff-669e68649d93c35ad69bd0724e11e6b181021954e5483076ad88604ee67d5fcbR389

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 13]
```diff
@@ -228,4 +237,116 @@ FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
...
+  }
+
+  static int32_t encodedSize() {
+    return 9;
+  }
+
+ private:
+  const int precision_;
+  const int scale_;
+  const int newPrecision_ = ShortDecimalType::kMaxPrecision;
```
Comment (skadilover) [inline velox/exec/prefixsort/PrefixSortEncoder.h:287]: This data member looks redundant

**Review by skadilover — COMMENTED** [review]
read this code again，here are some optimization suggestions:

1. Description

You can add some concise descriptions to express the intention of this MR, such as (perhaps not appropriate): 1. Add LongDecimalType support to PrefixSort (whether ShortDecimalType is planned to be supported and the reason why it cannot be supported in this PR) 2. Refer to the spark implementation, for precision - When scale <= ShortDecimalType::kMaxPrecision, decmal can be converted to int for processing

2.  necessary  code comments
Add necessary comments to public methods and methods whose intent is difficult to understand，for example：

canCastToInt

I think we need to add enough comments to explain why it could be work here ?

3. Ut with clear boundary conditions
For example ， ut that triggers the following logic： 
`
 if (FOLLY_UNLIKELY(!status.ok()))
`
And add necessary code comments to them to explain the triggering logic

**Comment by jinchengchenghh** [comment]
1. Short decimal TypeKind is INTEGER, it has been supported by you. And I have added short decimal to prefixsort fuzzer test. The implementation link is listed in the PR description.

**Comment by jinchengchenghh** [comment]
Do you have further comments? @skadilover

File: velox/exec/tests/PrefixSortTest.cpp [hunk 14]
```diff
@@ -219,13 +232,16 @@ TEST_F(PrefixSortTest, multipleKeys) {
...
 TEST_F(PrefixSortTest, fuzz) {
+  // Because for long decimal can cast to int cases, some edge values will be
+  // considered same, so we cannot test it in fuzzer test.
   std::vector<TypePtr> keyTypes = {
       INTEGER(),
       BOOLEAN(),
       TINYINT(),
       SMALLINT(),
       BIGINT(),
-      HUGEINT(),
```
Comment (skadilover) [inline velox/exec/tests/PrefixSortTest.cpp:228]: I think if this optimization introduces correctness issues, let's not introduce it.

**Comment by mbasmanova** [comment]
@jinchengchenghh There are CI failures. Please, take a look?

I see an unresolved comment from @skadilover :

> I understand the problem here, I think if "castToInt" introduces correctness issues, let's not introduce it，

We definitely do not want to introduce correctness bugs. Would you help clarify if there is a correctness issue that still needs resolving?

**Comment by skadilover** [comment]
> @jinchengchenghh There are CI failures. Please, take a look?
> 
> I see an unresolved comment from @skadilover :
> 
> > I understand the problem here, I think if "castToInt" introduces correctness issues, let's not introduce it，
> 
> We definitely do not want to introduce correctness bugs. Would you help clarify if there is a correctness issue that still needs resolving?

I think it`s still not resolving , the ut not change, hugeint is moved

File: velox/exec/prefixsort/PrefixSortEncoder.h (hunk 9 above)
Comment (skadilover) [reply velox/exec/prefixsort/PrefixSortEncoder.h:267]: @mbasmanova I think the diff with velox  is here: the case when !status.ok() may generate different result.

**Comment by mbasmanova** [comment]
CC: @FelixYBW @rui-mo

**Comment by jinchengchenghh** [comment]
Spark uses an approximately prefix sort like this PR implements, as this link shows,  https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala#L167, does Velox expects a concrete sort result? If so, I will move the castToInt optimization, only left encode int128 as int128. @mbasmanova

File: velox/exec/prefixsort/PrefixSortEncoder.h [hunk 15]
```diff
@@ -228,4 +237,128 @@ FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
...
+    }
+  }
+
+ private:
+  /// Spark implementation:
+  /// https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala#L167
+  /// Spark cast the decimal to int when the cast operation will not change its
+  /// actual value except some edge cases. For the value near to the maxium or
+  /// minimum value that this precision can represent, it use a appoximately
+  /// value INT64_MAX or INT64_MIN to represent.
```
Comment (rui-mo) [inline velox/exec/prefixsort/PrefixSortEncoder.h:358]: The prefix sort is under velox/exec, wondering if we shall not restrict the implementation to be Spark-specific. Perhaps we can check the semantics of Presto and handle the difference if there is.

**Comment by mbasmanova** [comment]
> Spark uses an approximately prefix sort 

@jinchengchenghh I'm not familiar with this. What is "approximately prefix sort"? Where can I read about it?

**Comment by jinchengchenghh** [comment]
This is spark implementation, it also uses a approximate value to compare and sort.
```
case dt: DecimalType if dt.precision - dt.scale <= Decimal.MAX_LONG_DIGITS =>
      val p = Decimal.MAX_LONG_DIGITS
      val s = p - (dt.precision - dt.scale)
      (raw) => {
        val value = raw.asInstanceOf[Decimal]
        if (value.changePrecision(p, s)) {
          value.toUnscaledLong
        } else if (value.toBigDecimal.signum < 0) {
          Long.MinValue
        } else {
          Long.MaxValue
        }
      }
```

And for string, it `Returns a 64-bit integer that can be used as the prefix used in sorting.`
```
  public long getPrefix() {
    return ByteArray.getPrefix(base, offset, numBytes);
  }
```
Further more, Spark converts all the data type prefix to long.
So it cannot sort the approximate value as its order.

In duckdb, it also extracts some bytes for string, but I'm not familiar with duckdb, I don't know how much it will extract.
https://github.com/duckdb/duckdb/blob/main/src/include/duckdb/common/radix.hpp#L46

I will try to read Presto sort code, if it is not same with Spark, I will change it by a config or move it to Spark folder. @mbasmanova

**Comment by jinchengchenghh** [comment]
The new progress, Spark use prefix sort(radix sort) in some conditions,
https://github.com/apache/spark/blob/branch-3.5/sql/core/src/main/scala/org/apache/spark/sql/execution/SortExec.scala#L81
```
val canUseRadixSort = enableRadixSort && sortOrder.length == 1 &&
      SortPrefixUtils.canSortFullyWithPrefix(boundSortExpression)
```
So Spark sort result for one partition is correct, further more, it supports sort global with the shuffle data redistribution.
https://github.com/apache/spark/blob/branch-3.5/sql/core/src/main/scala/org/apache/spark/sql/execution/SortExec.scala#L42

And in SpillMerger, it will firstly check whether the result prefix is equal, if equal, compare the data by `recordComparator`.

In Velox, we don't check the actual data record, only compare the prefix, so I will serialize LongDecimal to int128_t in that case. @mbasmanova @skadilover 
CC @rui-mo

**Comment by jinchengchenghh** [comment]
Can you help review again? Now it is a very simple version. @skadilover  Thanks!

**Comment by jinchengchenghh** [comment]
Can you help take a look? Thanks! @mbasmanova

File: velox/exec/tests/PrefixSortTest.cpp (hunk 14 above)
Comment (jinchengchenghh) [reply velox/exec/tests/PrefixSortTest.cpp:228]: Now I remove the cast to int, only keeps encode int128 itself.

**Review by xiaoxmeng — APPROVED** [review]
@jinchengchenghh LGTM. Thanks for the update!

File: velox/exec/tests/PrefixSortTest.cpp [hunk 16]
```diff
@@ -147,6 +147,12 @@ TEST_F(PrefixSortTest, singleKey) {
       makeFlatVector<int64_t>({5, 4, 3, 2, 1}),
       makeFlatVector<int32_t>({5, 4, 3, 2, 1}),
       makeFlatVector<int16_t>({5, 4, 3, 2, 1}),
+      makeFlatVector<int128_t>(
+          {5,
+           HugeInt::build(20, 400),
+           HugeInt::build(20, 401),
+           HugeInt::build(20, 402),
+           HugeInt::build(-20, 401)}),
```
Comment (majetideepak) [inline velox/exec/tests/PrefixSortTest.cpp:155]: @skadilover Can we use `HugeInt::parse` instead of HugeInt::build? 
HugeInt::build only supports unsigned integers even though HugeInt should support signed.
I will add a PR to fix these semantics.

File: velox/exec/tests/PrefixSortTest.cpp [hunk 17]
```diff
@@ -186,6 +192,12 @@ TEST_F(PrefixSortTest, singleKeyWithNulls) {
       makeNullableFlatVector<int64_t>({5, 4, std::nullopt, 2, 1}),
       makeNullableFlatVector<int32_t>({5, 4, std::nullopt, 2, 1}),
       makeNullableFlatVector<int16_t>({5, 4, std::nullopt, 2, 1}),
+      makeNullableFlatVector<int128_t>(
+          {5,
+           HugeInt::build(20, 400),
```
Comment (xiaoxmeng) [inline velox/exec/tests/PrefixSortTest.cpp:197]: How about these?
